paddle/framework/lod_tensor.h

```diff
@@ -25,20 +26,58 @@ limitations under the License. */
 #include "paddle/framework/ddim.h"
 #include "paddle/framework/tensor.h"
 #include "paddle/framework/tensor_util.h"
+#include "paddle/memory/memcpy.h"
 #include "paddle/platform/enforce.h"
 #include "paddle/platform/place.h"
 
 namespace paddle {
 namespace framework {
 
-#ifndef PADDLE_WITH_CUDA
-template <typename T>
-using Vector = std::vector<T>;
-#else
+/**
+ * @brief Vector support both cpu and gpu.
+ * cpu vector lifetime is same with Vector
+ * gpu vector is lazily malloc and modified.
+ */
+
 template <typename T>
-using Vector = thrust::host_vector<
-    T, thrust::system::cuda::experimental::pinned_allocator<T>>;
-#endif
+class Vector : public std::vector<T> {
+ public:
+  using std::vector<T>::vector;
+
+  Vector() {}
+  Vector(const std::vector<T>& v) : std::vector<T>(v) {}  // NOLINT
+
+  virtual ~Vector() {
+    if (cuda_ptr_ != nullptr) {
+      memory::Free<platform::CUDAPlace>(place_, static_cast<void*>(cuda_ptr_));
+    }
+  }
+
+  T* cuda_data() {
+    CopyToCUDA();
+    PADDLE_ENFORCE_NOT_NULL(
+        cuda_ptr_, "No data or Insufficient CUDA memory to allocation");
+    return static_cast<T*>(cuda_ptr_);
+  }
+
+  const T* cuda_data() const {
+    CopyToCUDA();
+    PADDLE_ENFORCE_NOT_NULL(
+        cuda_ptr_, "No data or Insufficient CUDA memory to allocation");
+    return static_cast<const T*>(cuda_ptr_);
+  }
+
+  void CopyToCUDA();
+
+  void CopyFromCUDA();
+
+  void CopyToPeer();
```

Comment:
There should be a flag for data has been at GPU memory, and do not copy it again.

We should overload all method of std::vector and maintain this flag.